src/main/java/org/disq_bio/disq/impl/formats/cram/CramSource.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.broadcast.Broadcast;
+import org.disq_bio.disq.impl.file.FileSystemWrapper;
+import org.disq_bio.disq.impl.file.PathChunk;
+import org.disq_bio.disq.impl.file.PathSplit;
+import org.disq_bio.disq.impl.file.PathSplitSource;
+import org.disq_bio.disq.impl.formats.SerializableHadoopConfiguration;
+import org.disq_bio.disq.impl.formats.sam.AbstractBinarySamSource;
+import org.disq_bio.disq.impl.formats.sam.SamFormat;
+
+public class CramSource extends AbstractBinarySamSource implements Serializable {
+
+  private final PathSplitSource pathSplitSource;
+
+  public CramSource(FileSystemWrapper fileSystemWrapper) {
+    super(fileSystemWrapper);
+    this.pathSplitSource = new PathSplitSource(fileSystemWrapper);
+  }
+
+  @Override
+  public SamFormat getSamFormat() {
+    return SamFormat.CRAM;
+  }
+
+  @Override
+  protected JavaRDD<PathChunk> getPathChunks(
+      JavaSparkContext jsc,
+      String path,
+      int splitSize,
+      ValidationStringency stringency,
+      String referenceSourcePath)
+      throws IOException {
+
+    final Configuration conf = jsc.hadoopConfiguration();
+
+    // store paths (not full URIs) to avoid differences in scheme - this could be improved
+    Map<String, NavigableSet<Long>> pathToContainerOffsets = new LinkedHashMap<>();
+    if (fileSystemWrapper.isDirectory(conf, path)) {
+      List<String> paths =
+          fileSystemWrapper
+              .listDirectory(conf, path)
+              .stream()
+              .filter(SamFormat.CRAM::fileMatches)
+              .collect(Collectors.toList());
+      for (String p : paths) {
+        long cramFileLength = fileSystemWrapper.getFileLength(conf, p);
+        NavigableSet<Long> containerOffsets = getContainerOffsetsFromIndex(conf, p, cramFileLength);
+        String normPath = URI.create(fileSystemWrapper.normalize(conf, p)).getPath();
+        pathToContainerOffsets.put(normPath, containerOffsets);
+      }
+    } else {
+      long cramFileLength = fileSystemWrapper.getFileLength(conf, path);
+      NavigableSet<Long> containerOffsets =
+          getContainerOffsetsFromIndex(conf, path, cramFileLength);
+      String normPath = URI.create(fileSystemWrapper.normalize(conf, path)).getPath();
+      pathToContainerOffsets.put(normPath, containerOffsets);
+    }
+    Broadcast<Map<String, NavigableSet<Long>>> containerOffsetsBroadcast =
+        jsc.broadcast(pathToContainerOffsets);
+
+    SerializableHadoopConfiguration confSer =
+        new SerializableHadoopConfiguration(jsc.hadoopConfiguration());
+
+    return pathSplitSource
```

Comment:
Clarifying question, when you split crams into partitions, they are being split so each cram container is held in its entirety within a partition? Does what happens in degenerate cases with very large cram containers that number fewer than the target number of partitions for splitting?

Reply:
CRAM containers are not split, so each one is read entirely by exactly one partition. In the case of a container that is bigger than the file split size some partitions will be empty.